### PR TITLE
[#176595075] Add HTTPS fetch for API and Bonus client

### PR DESCRIPTION
**Pull request by BurnedMarshal**
#### List of Changes
- API Client and Bonus Client can use HTTPS and HTTP fetch based on provided url

#### Motivation and Context
When `API_URL` and `BONUS_API_URL` use https protocol a specific https fetch is needed. Otherwise an error occurs 
```
Internal server error: Exception while calling upstream API (likely a timeout).
[TypeError [ERR_INVALID_PROTOCOL]: Protocol "https:" not supported. Expected "http:"]
```

#### How Has This Been Tested?
- Tested locally with `io-mock` using a new configuration proposed in pagopa/io-mock#32

#### Types of changes

- [ ] Chore
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -333,15 +341,15 @@ log.info("Session token duration set to %s seconds", tokenDurationSecs);
...
+const simpleHttpsApiFetch = agent.getHttpsFetch(process.env);
 
 // Register a factory service to create PagoPA client.
 const pagoPAApiUrlProd = getRequiredENVVar("PAGOPA_API_URL_PROD");
 const pagoPAApiUrlTest = getRequiredENVVar("PAGOPA_API_URL_TEST");
 export const PAGOPA_CLIENT = new PagoPAClientFactory(
   pagoPAApiUrlProd,
   pagoPAApiUrlTest,
-  httpsApiFetch
+  simpleHttpsApiFetch
```
Comment (balanza) [inline src/config.ts:352]: As we are here, should we implement this for both cases as well? (a bit complex in this case, I see)

Comment (BurnedMarshal) [reply src/config.ts:352]: Maybe we can add a constraint on URLs in this case with a `PetternString` type check
```typescript
const HttpsUrl = PatternString("^(https:\/\/).+$");
type HttpsUrl = t.TypeOf<typeof HttpsUrl>;

// Register a factory service to create PagoPA client.
const pagoPAApiUrlProd = HttpsUrl.decode(
  process.env.PAGOPA_API_URL_PROD
).getOrElseL(() => process.exit(1));
const pagoPAApiUrlTest = HttpsUrl.decode(
  process.env.PAGOPA_API_URL_PROD
).getOrElseL(() => process.exit(1));
```

Comment (balanza) [reply src/config.ts:356]: So that we enforce such url to be https only. Do we want this?

Comment (gunzip) [reply src/config.ts:356]: https://github.com/pagopa/io-ts-commons/blob/master/src/url.ts#L31

Comment (BurnedMarshal) [reply src/config.ts:356]: This check is evaluated at startup and prevents errors at runtime, sometimes this type of enforcement could be useful.

**Review by gunzip — COMMENTED** [review]
how does this plays against https://github.com/pagopa/io-ts-commons/pull/348 ?

**Comment by BurnedMarshal** [comment]
> how does this plays against [pagopa/io-ts-commons#348](https://github.com/pagopa/io-ts-commons/pull/348) ?

When the changes in `io-ts-commons` will be available this PR can change into package upgrade and use of generic `getFetch ` in the meantime I change to draft this PR.

**Comment by balanza** [comment]
> > how does this plays against [pagopa/io-ts-commons#348](https://github.com/pagopa/io-ts-commons/pull/348) ?
> 
> When the changes in `io-ts-commons` will be available this PR can change into package upgrade and use of generic `getFetch ` in the meantime I change to draft this PR.

Published https://www.npmjs.com/package/@pagopa/ts-commons/v/9.1.0

**Comment by AleDore** [comment]
Can we perform a rebase @BurnedMarshal ?

**Comment by BurnedMarshal** [comment]
> Can we perform a rebase @BurnedMarshal ?

@AleDore rebase done

**Review by AleDore — APPROVED** [review]
LGTM. As a general comment we should deploy this new version in a progressive way in order to check API client's behaviour

**Pull request merged**